src/main/java/gift/service/OptionServiceImpl.java

```diff
@@ -0,0 +1,111 @@
+package gift.service;
+
+import gift.dto.CreateOptionRequestDto;
+import gift.dto.OptionResponseDto;
+import gift.dto.PurchaseOptionRequestDto;
+import gift.dto.UpdateOptionQuantityRequestDto;
+import gift.entity.Option;
+import gift.entity.Product;
+import gift.exception.CustomException;
+import gift.exception.ErrorCode;
+import gift.repository.OptionRepository;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+public class OptionServiceImpl implements OptionService {
+
+    private final OptionRepository optionRepository;
+
+    private final ProductService productService;
+
+    public OptionServiceImpl(OptionRepository optionRepository, ProductService productService) {
+        this.optionRepository = optionRepository;
+        this.productService = productService;
+    }
+
+    @Override
+    public List<OptionResponseDto> findProductOptionById(Long id) {
+        List<Option> options = findOptionsByProductIdOrElseThrow(id);
+        return toOptionResponseDtoList(options);
+    }
+
+    @Override
+    @Transactional
+    public OptionResponseDto createOption(CreateOptionRequestDto requestDto, Long productId) {
+        checkDuplicateOption(productId, requestDto.name());
+        Option newOption = new Option(requestDto.name(), requestDto.quantity(), null);
+        Product product = productService.findProductByIdOrElseThrow(productId);
+        newOption.setProduct(product);
+        Option savedOption = optionRepository.save(newOption);
+        return new OptionResponseDto(savedOption.getId(), savedOption.getName(),
+                savedOption.getQuantity());
+    }
+
+    @Override
+    @Transactional
+    public OptionResponseDto setOptionQuantity(
+            Long id,
+            Long optionId,
+            UpdateOptionQuantityRequestDto requestDto) {
+        Option option = findOptionByProductIdAndOptionIdOrElseThrow(id, optionId);
+        option.changeQuantity(requestDto.quantity());
+        Option updatedOption = findOptionByProductIdAndOptionIdOrElseThrow(id, optionId);
+        return new OptionResponseDto(updatedOption.getId(), updatedOption.getName(),
+                updatedOption.getQuantity());
+    }
+
+    @Override
+    @Transactional
+    public OptionResponseDto purchaseOption(
+            Long id,
+            Long optionId,
+            PurchaseOptionRequestDto requestDto) {
+        Option option = findOptionByProductIdAndOptionIdOrElseThrow(id, optionId);
+        if (requestDto.quantity() > option.getQuantity()) {
```

Comment:
객체지향 생활 체조 원칙 
> 규칙 9: getter/setter/프로퍼티를 쓰지 않는다.

getter를 제거해보는게 어떨까요?

Reply:
엔티티에 값을 비교하는 메서드를 추가할 수 있을 거 같습니다.
```
public boolean hasEnoughQuantity(Long quantity) {
    return this.quantity >= quantity;
}
```
```suggestion
        if (!option.hasEnoughQuantity(requestDto.quantity())) {
```